goalie/adjoint.py

```diff
@@ -221,6 +221,77 @@ def _create_solutions(self):
         """
         self._solutions = AdjointSolutionData(self.time_partition, self.function_spaces)
 
+    def _extract_adjoint_solutions(self, field, i, solve_blocks, get_adj_values=False):
+        """
+        Extract adjoint solutions from the tape for a given subinterval.
+
+        :arg field: field to extract
+        :type field: :class:`str`
+        :arg i: subinterval index
+        :type i: :class:`int`
+        :kwarg get_adj_values: if ``True``, adjoint actions are also returned at exported
+            timesteps
+        :type get_adj_values: :class:`bool`
+        """
+        # TODO #125: Support get_adj_values in AdjointSolutionData
+        tp = self.time_partition
+        solutions = self.solutions.extract(layout="field")[field]
+        if get_adj_values and "adj_value" not in solutions:
+            fs = self.function_spaces[field]
+            self.solutions.extract(layout="field")[field]["adj_value"] = []
+            for i, fs in enumerate(self.function_spaces[field]):
+                self.solutions.extract(layout="field")[field]["adj_value"].append(
+                    [
+                        firedrake.Cofunction(fs.dual(), name=f"{field}_adj_value")
+                        for j in range(tp.num_exports_per_subinterval[i] - 1)
+                    ]
+                )
+
+        # Update adjoint solution data based on block dependencies and outputs
+        stride = tp.num_timesteps_per_export[i]
+        for j, block in enumerate(reversed(solve_blocks[::-stride])):
+            # Current adjoint solution is determined from the adj_sol attribute
+            if block.adj_sol is not None:
+                solutions.adjoint[i][j].assign(block.adj_sol)
+
+            # Adjoint action comes from dependencies
+            dep = self._dependency(field, i, block)
+            if get_adj_values and dep is not None:
+                solutions.adj_value[i][j].assign(dep.adj_value)
+
+            # The adjoint solution at the 'next' timestep is determined from the
+            # adj_sol attribute of the next solve block
+            num_solve_blocks = len(solve_blocks)
+            steady = self.steady or len(self) == num_solve_blocks == 1
+            if not steady:
+                if (j + 1) * stride < num_solve_blocks:
+                    if solve_blocks[(j + 1) * stride].adj_sol is not None:
+                        solutions.adjoint_next[i][j].assign(
+                            solve_blocks[(j + 1) * stride].adj_sol
+                        )
+                elif (j + 1) * stride > num_solve_blocks:
+                    raise IndexError(
+                        "Cannot extract solve block"
+                        f" {(j + 1) * stride} > {num_solve_blocks}."
+                    )
```

Comment:
Is this necessary despite the existing checks in `time_partition.py`? I.e. can it happen that `firedrake.adjoint_utils.solving.get_solve_blocks` return less solve blocks than we anticipate? If so, I'd rather add this check to `MeshSeq.get_solve_blocks` and remove it here.